### PR TITLE
Java. Page Factory enhancement.

### PR DESCRIPTION
**Pull request by TikhomirovSergey**
- Ability to use not only Webdriver to populate page object fields via method

``` java
public static <T> T initElements(SearchContext context, Class<T> pageClassToProxy).
```

This change was done in order to simplify page object instantiation and population when there is the necessity to implement something that describes widget or repeatable element set nested in a root element. So if there is a class with constructors like follows:

``` java
public PageObject(Search context)

//or

public PageObject(WebElement rootElement)
```

and there is no necessity to implement customized decorating engines then it is possible to populete its WebElement-fields this way (only one step):

``` java
PageObject po = PageFactory.initElements(driver, PageObject,class) //then PageObject.class has 
//the default constructor or constructor with a SearchContext- or WebDriver- parameter


//or 
PageObject po = PageFactory.initElements(element, PageObject,class) //then PageObject.class has 
//the default constructor or constructor with a SearchContext- or WebElement- parameter

//and so on... 
```
- the additional checking of field modifiers before it will be populated. 

Field that is going to be filled should not  be static and final. I found these issues some time ago. I tried to implement page object with final fields that weren't annotated by @FindBy's. Also I found that it tries to set proxy-WebElement as a value of a static WebElement-field. In both cases I caught **IllegalModifierException**. Actually I think that both issues are minor but it would be good if whey were fixed.

---
This change is [Reviewable](https://reviewable.io/reviews/seleniumhq/selenium/415)

### CODE REVIEW

**Comment by opensource21** [comment]
Hi I made some similar [change at selophane](https://github.com/elisarver/selophane/commit/84f09009cfdc62e5a135183d529be09dd94097c8). How ever the last days I found a use case where it's necessary to have the webdriver. So I changed it back. The background is, that if you create something like a button it would be great if there was a wait till a click on the button is finished. Therefor you need a WebDriver. 
You can see my current solution in a [pull request](https://github.com/elisarver/selophane/pull/23)

**Comment by TikhomirovSergey** [comment]
@opensource21 
There is nothing prevents to use Webdriver if this change will be accepted. Actually it is possible to use WebElement for the subelements populating now:

```java
//excessive actions
WidgetObject wo = new WidgetObject(/*refToElement*/);
DefaultFieldDecorator dfd = new DefaultFieldDecorator(
new DefaultElementLocatorFactory(refToElement)); 
//excessive actions are finished
PageFactory.initElements(dfd, wo);
```
This should be usefull for cases and apps/services like... Google (just for example).  The search form is the page by itself and the widget inside some element at Drive of GMail page. So if it is not necessary to made up and implement users customized decorator then this widget object will be created and populated in one action.

**Comment by opensource21** [comment]
@TikhomirovSergey Yes you are right! Sorry for the trouble and thanks for the hints.

**Comment by ddavison** [comment]
there appears to be an issue with [this travis build](https://travis-ci.org/SeleniumHQ/selenium/builds/82942245).

no ruby changes were detected, yet still failed

**Comment by titusfortner** [comment]
I'm not sure why it ran in the first place, but it looks like I might have changed something in the specs. I'll take a look.

**Comment by TikhomirovSergey** [comment]
@mykola-mokhnach @shs96c
It seems that PR has been actualized. 
Could you rewiew this PR?

Also I don't know the reason why tests are not passed

File: java/client/src/org/openqa/selenium/support/PageFactory.java [hunk 1]
```diff
@@ -122,19 +126,25 @@ private static void proxyFields(FieldDecorator decorator, Object page, Class<?>
...
+
+          Class<?>[] parameterTypes = c.getParameterTypes();
+          if (parameterTypes.length != 1)
+            continue;
+
+          Class<?> parameterClazz = parameterTypes[0];
+          if (!parameterClazz.isAssignableFrom(context.getClass()))
+            continue;
+          c.setAccessible(true);
+          return (T) c.newInstance(context);
```
Comment (mykola-mokhnach) [inline java/client/src/org/openqa/selenium/support/PageFactory.java]: _pageClassToProxy.cast_ should address unchecked class cast warning

File: java/client/src/org/openqa/selenium/support/PageFactory.java [hunk 2]
```diff
@@ -122,19 +126,25 @@ private static void proxyFields(FieldDecorator decorator, Object page, Class<?>
...
+  @SuppressWarnings("unchecked")
+  private static <T> T instantiatePage(SearchContext context, Class<T> pageClassToProxy) {
     try {
-      try {
-        Constructor<T> constructor = pageClassToProxy.getConstructor(WebDriver.class);
-        return constructor.newInstance(driver);
-      } catch (NoSuchMethodException e) {
-        return pageClassToProxy.newInstance();
+      Constructor<?>[] availableConstructors = pageClassToProxy.getDeclaredConstructors();
+      for (Constructor<?> c: availableConstructors){
```
Comment (mykola-mokhnach) [inline java/client/src/org/openqa/selenium/support/PageFactory.java]: a space is missing

File: java/client/src/org/openqa/selenium/support/PageFactory.java [hunk 3]
```diff
@@ -122,19 +126,25 @@ private static void proxyFields(FieldDecorator decorator, Object page, Class<?>
...
+
+          Class<?>[] parameterTypes = c.getParameterTypes();
+          if (parameterTypes.length != 1)
+            continue;
```
Comment (mykola-mokhnach) [inline java/client/src/org/openqa/selenium/support/PageFactory.java]: it is a good practice to always enclose condition block into curly brackets, even if there is only one operator

File: java/client/src/org/openqa/selenium/support/pagefactory/internal/LocatingElementHandler.java [hunk 4]
```diff
@@ -33,13 +33,15 @@ public LocatingElementHandler(ElementLocator locator) {
   }
 
   public Object invoke(Object object, Method method, Object[] objects) throws Throwable {
+
+    if ("toString".equals(method.getName())) {
+      return "Proxy element for: " + locator.toString();
```
Comment (mykola-mokhnach) [inline java/client/src/org/openqa/selenium/support/pagefactory/internal/LocatingElementHandler.java]: String.format might be a better choice here

**Comment by TikhomirovSergey** [comment]
@mykola-mokhnach @shs96c I have updated this PR. Please look at this once again.

File: java/client/src/org/openqa/selenium/support/PageFactory.java [hunk 5]
```diff
@@ -110,6 +110,10 @@ public static void initElements(FieldDecorator decorator, Object page) {
   private static void proxyFields(FieldDecorator decorator, Object page, Class<?> proxyIn) {
     Field[] fields = proxyIn.getDeclaredFields();
     for (Field field : fields) {
+      int modifiers = field.getModifiers();
+      if (Modifier.isFinal(modifiers) || Modifier.isStatic(modifiers))
```
Comment (mykola-mokhnach) [inline java/client/src/org/openqa/selenium/support/PageFactory.java]: curly brackets would be good here as well

File: java/client/src/org/openqa/selenium/support/PageFactory.java [hunk 6]
```diff
@@ -122,19 +126,26 @@ private static void proxyFields(FieldDecorator decorator, Object page, Class<?>
...
+          if (parameterTypes.length != 1) {
+            continue;
+          }
+
+          Class<?> parameterClazz = parameterTypes[0];
+          if (!parameterClazz.isAssignableFrom(context.getClass())) {
+            continue;
+          }
+          c.setAccessible(true);
+          return (T) c.newInstance(context);
```
Comment (mykola-mokhnach) [inline java/client/src/org/openqa/selenium/support/PageFactory.java]: did pageClassToProxy.cast fail?

Comment (TikhomirovSergey) [reply java/client/src/org/openqa/selenium/support/PageFactory.java]: Ok. I can improve it later.

**Comment by diemol** [comment]
Apologies for the poor handling of this PR.

In general, we are discouraging the use of `PageFactory`, but in any case, lots of people use, so making these types of changes would break backward compatibility. Closing this PR.

**Pull request closed without merging**